mtcpclient/calculate.go

```diff
@@ -17,6 +17,12 @@ type metricsCollectionStats struct {
 }
 
 func (gc GroupOfConnections) calculateMetricsReport(status tcpclient.ConnectionStatus) (mr metricsCollectionStats) {
+	// TODO: There's something i don't like... initiatlizing values in the first loop, and the standard deviation
+	//  requiring an extra pass considering all items... i'd move initialization out of the loop, and maybe iterate
+	//  over a filtered list rather than several loops over the original one, and maybe use specific generic functions...
+	//  requires further thinking in any case
+	// TODO: (2) rather than calculate a metrics report with a filter on the connection status, lets just promote splitting
```

Comment:
It's an option, but is this more efficient? I mean you are doing a first loop to filter, store a new structure with the filtered list and going through it to get the metrics.

Reply:
Well, if all decisions are made based on efficiency, probably we should be programming in assembler... of course, you can try to reduce "iterations" and do everything you need on each item... but this probably means doing too much staff in the same function, and this is a very bad practice unless, as you point, is critical in terms of performance. In fact, we've already broken this "practice" by moving std deviation outside of this function (and assuming more than one iteration then...). Making an specific assessment for this case... a reporting function that runs at the very end of the program (not part of the "main" program functionality, that is doing the tcp conn mgmt), that works with an structure that won't have more than 1k-10k elements... do you really think is worth to sacrifice readability / simplicity in favour of avoiding an extra loop over this structure? In fact, making a first loop to pick the elements that are actually important will minimise the number of items you will consider in future loops... So my bet on removing methods tight to a "group of connections" that use filters + do the logic against the filtered items, versus methods that do filtering and different methods that work against a group of connections (that probably have been already filtered) is extremely clear and I'm pretty sure you will notice the difference from a "clean-code" perspective once you see this implemented :)

Reply:
Yup, I agree with you... we should code in assembler (just kidding)
I agree with you that cleanliness is more important writing code than performance (if it's not critical), and splitting responsibilities makes everything easier to maintain an evolve.
I was just wondering when to take the step in each direction...